### PR TITLE
expression, executor, plan: rewrite builtin function replace.

### PR DESCRIPTION
**Pull request by SteveZhangBit**

### CODE REVIEW

File: expression/builtin_string.go [hunk 1]
```diff
@@ -585,46 +585,53 @@ type replaceFunctionClass struct {
...
+	}
+	return charLen
+}
+
 type builtinReplaceSig struct {
-	baseBuiltinFunc
+	baseStringBuiltinFunc
 }
 
 // eval evals a builtinReplaceSig.
```
Comment (zz-jason) [inline expression/builtin_string.go]: update comment, change `eval` to `evalString`

File: expression/builtin_string.go [hunk 2]
```diff
@@ -585,46 +585,53 @@ type replaceFunctionClass struct {
 }
 
 func (c *replaceFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
-	sig := &builtinReplaceSig{newBaseBuiltinFunc(args, ctx)}
+	bf, err := newBaseBuiltinFuncWithTp(args, ctx, tpString, tpString, tpString, tpString)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	bf.tp.Flen = c.fixLength(args)
```
Comment (zz-jason) [inline expression/builtin_string.go]: flag/charset/collation of `bf.tp` should be binary as long as one of the parameters is a binary string

File: expression/builtin_string.go [hunk 3]
```diff
@@ -585,46 +585,59 @@ type replaceFunctionClass struct {
...
+			types.SetBinChsClnFlag(bf.tp)
+			break
+		}
+	}
+	sig := &builtinReplaceSig{baseStringBuiltinFunc{bf}}
 	return sig.setSelf(sig), errors.Trace(c.verifyArgs(args))
 }
 
+// fixLength calculate the Flen of the return type.
+// See https://github.com/mysql/mysql-server/blob/5.7/sql/item_strfunc.cc#L1478
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: remove this comment `See ...`

**Comment by SteveZhangBit** [comment]
@XuHuaiyu 
fixed

**Pull request merged**